### PR TITLE
Add some simple rules for common situations

### LINKED ISSUE #79: Explain that you are allowed to have GPL code as dependency
(for Apache 2.0 code)

### PR DESCRIPTION
**Pull request by LourensVeen**
Fixes #79, and hopefully saves people some time :).

### CODE REVIEW

File: intellectual_property/some_rules.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+- BSD 2-clause
+- BSD 3-clause
+- Apache License version 2.0
+- GNU Lesser General Public License v2 or later
+- GNU General Public License v2 or later
+
+and any other licenses, including "for academic use only" and similar statements, are not okay.
+
+
+*Rationale: For all but the copyleft licenses in this list, the GNU GPL v2 or later, this is clearly allowed, regardless of whether the program is considered a derivative work.
```
Comment (nielsdrost) [inline intellectual_property/some_rules.md]: Strongly disagree with this rationale. Seems like a loophole to me. Exploiting legal loopholes is not the way to advance science.

**Comment by nielsdrost** [comment]
Strongly disagree with the licensing text here. If you use GPL code your code should be GPL as well. The author put GPL on the code for a reason: for it, or any of its derivatives, not to ever become closed source unless with the authors explicit permission. There are plenty of alternative licenses to choose from if this was not the intention.

**Comment by LourensVeen** [comment]
It seems from our off-line conversation and your comment above that we disagree pretty fundamentally in how we interpret the GPL and the intent behind it. I will try to explain better why I think it is okay to distribute the source code of a program that uses a GPL library under the Apache License, as long as the works are separable (which they clearly are if they're distributed separately). I'll cover both the moral and the legal standpoints.

# Moral aspects
The preamble to the GPL (version 3, but the one of version 2 is nearly identical), says:

> When we speak of free software, we are referring to freedom, not price. Our General Public Licenses are designed to make sure that you have the freedom to distribute copies of free software (and charge for this service if you wish), that you receive source code or can get it if you want it, that you can change the software or use pieces of it in new free programs; and that you know you can do these things.

Note that this is a message from the owner of the library to us, and to everyone. I think it is clear that by us distributing our source code under the Apache License, everyone still has the freedom to distribute copies of the library, and also to clone our repository, and combine them and distribute the resulting work, for a charge or not. 

Everyone can also get the full source code to everything if they want it, and change any of the pieces, use them in new free programs, and as long as every piece of code is clearly marked with respect to the copyright ownership and licensing, everyone knows that they can do these things.

(There is a technical issue with respect to compatibility of the Apache License version 2 and the GNU GPL version 2, more on that below, as it's a legal thing, not a moral one, as shown by the fact that it's been fixed in the GPLv3.)

But there is more in that preamble:

> To protect your rights, we need to make restrictions that forbid anyone to deny you these rights or to ask you to surrender the rights. These restrictions translate to certain responsibilities for you if you distribute copies of the software, or if you modify it.

> For example, if you distribute copies of such a program, whether gratis or for a fee, you must give the recipients all the rights that you have. You must make sure that they, too, receive or can get the source code. And you must show them these terms so they know their rights.

We are not distributing the library, but we are distributing our own code. We are distributing it gratis (although for a fee would also be fine), and we give the recipients all the same rights that we are getting from the makers of the GPLled library, and then some. They can get or receive the source code, too, and we put a license notice on it to that effect so that everyone knows.

In the past, there have been several propaganda campaigns against Free Software in general, and the GNU GPL in particular. Companies making proprietary software, and benefiting from vendor lock-in and the resulting lack of competition, felt threatened by what was essentially a consumer rights movement, especially when the idea gained traction and people started making competing software that did not lock in users.

They started calling copyleft licenses (so-called because they precisely reverse the restrictions copyright places on creative works) anti-commercial, "viral", communist, and even un-American. Which perhaps they are, if you narrow down "commercial" to the business model of selling copies and prohibiting copying, if you think you should have the unfettered right to use other people's work in your proprietary software without compensation or reciprocation, if you consider replacing a highly anti-competitive business model with a free market for programming services to be communism, or if you consider anything that inconveniences a US company to be un-American.

While this propaganda has now died down somewhat, some damage has been done. There are still people who think that copyleft software cannot be run for commercial purposes, that any changes to it must be published, or that using copyleft software means that you have to publish all your other software as Free Software. And there will be some people out there who license their code under the GPL with those expectations.

But since it is impossible to know how exactly every single licensor of source code understands the GPL, I would argue that our only reasonable course of action is to take the license at face value, and base our actions on what it plainly says. The alternative, to base it on what people think who have been misled by parties hostile to Free Software, to me is simply insulting to the FSF.

From the preamble above, I don't see how publishing the contents of our repository under the Apache License contradicts the goals of the GPL. The main difference between us licensing our code under the Apache License as opposed to under the GPL would be in the following scenario: Someone clones our repository, modifies our source code to remove the dependency on the GPLled library, and uses the resulting code in their software, which they then distribute under a proprietary license.

Should the makers of the GPLled library be able to prevent this? I would argue that this is between us and the makers of the proprietary software, and that the makers of the GPLled library don't get a say, just because some previous version of our code used their code.

# Legal aspects
Of course, we have to do our legal due diligence as well. So let's have a look at the legal side of things. Should the GPL apply to our source code? The GPL version 2 says about applicability:

> This License applies to any program or other work which contains a notice placed by the copyright holder saying it may be distributed under the terms of this General Public License. The "Program", below, refers to any such program or work, and a "work based on the Program" means either the Program or any derivative work under copyright law: that is to say, a work containing the Program or a portion of it, either verbatim or with modifications and/or translated into another language.

Farther down, in section 2, it says:

> These requirements apply to the modified work as a whole. If identifiable sections of that work are not derived from the Program, and can be reasonably considered independent and separate works in themselves, then this License, and its terms, do not apply to those sections when you distribute them as separate works. [...] Thus, it is not the intent of this section to claim rights or contest your rights to work written entirely by you; rather, the intent is to exercise the right to control the distribution of derivative or collective works based on the Program.

Note that our source code repository does not contain the program or a portion of it, modified or not. Copyright law does not specify exactly what a derivative work is, and there's no case law with respect to software, but the definition in the license itself above clearly excludes our source code.

The GNU GPL version 3 says:

> To “modify” a work means to copy from or adapt all or part of the work in a fashion requiring copyright permission, other than the making of an exact copy. The resulting work is called a “modified version” of the earlier work or a work “based on” the earlier work.

> A “covered work” means either the unmodified Program or a work based on the Program.

Again, our source code does not copy from or adapt the library, it's just intended to be used with the library. So again that would mean that we are free to license our source code in whichever way we want. (Although morally I'd argue that it should be a Free license, see above, but we're doing that.)

And it has a further clause that very explicitly gives permission for distributing our source under a license that is less restricted than the GPLv3:

> “Additional permissions” are terms that supplement the terms of this License by making exceptions from one or more of its conditions. [...] If additional permissions apply only to part of the Program, that part may be used separately under those permissions, but the entire Program remains governed by this License without regard to the additional permissions.

I therefore argue that it is both morally and legally okay to license the source code in our repository under the Apache License, even if it depends on a GPLled library somewhere else.

# ALv2 vs. GPLv2

Finally there's one more thing to note: the Apache License version 2 is incompatible with the GNU GPL version 2. More specifically, the patent defense clause in the Apache License is more restrictive than the one in the GPLv2, and the GPL forbids adding more restrictions to a combined work. This is purely a technical point, and the Free Software Foundation added some additional language to the GPL version 3 that explicitly allows the Apache License's patent clause, but they could not retroactively change the GPL version 2.

This means that the combination of an Apache licensed work with a GPLv2 licensed work cannot be redistributed, which is inconvenient. In such cases, we should probably license our work under both the Apache license and the GPL version 2, so that any combined works can be distributed as GPLv2, and anyone using only our source can still do so under the Apache license. I'll add some text to clarify this.

**Comment by nielsdrost** [comment]
He Lourens,

Thanks for the long reply! You are definitely putting a lot of thought into this :-)

Unfortunately (though probably not unexpectedly), I still disagree. My point is that most, if not all, authors use GPL with the _intention_ of making sure all works that "build on" their work to also be GPL, and this is how everyone I know is both using and explaining the GPL. Case in point, the [licence selection page by github](https://choosealicense.com/licenses/gpl-3.0/) puts it like this:


> GNU GPLv3 
> Permissions of this strong copyleft license are conditioned on making available complete source code of licensed works and modifications, which include larger works using a licensed work, under the same license. Copyright and license notices must be preserved. Contributors provide an express grant of patent rights. 

So, even if when reading the GPL carefully we do not (fully) agree with the summary above, this is the widely accepted interpretation of this license. We should respect the author's choice for this license, and not build on their work in a way they very likely did not want to happen.

If they are ok with this, why would they not use LGPL or Apache?

**Comment by LourensVeen** [comment]
Just ran into this, from the FSF: http://www.softwarefreedom.org/resources/2007/gpl-non-gpl-collaboration.html

Haven't had time to read it yet, but will do so and revise accordingly, if needed.

**Comment by sonjageorgievska** [comment]
Thanks for the link, Lourens.

I did not read the discussion, too long indeed, but IMHO there should be a way to licence your software as Apache if it is calling external GPL libraries.

For example, if someone forks your software and replaces the GPL dependencies with Apache dependencies, then there would be no trace of the original GPL libraries in the new fork at all. He/she should be able to assign an Apache licence to the new tool.

There was a case like this in that workshop that I mentioned,  and the conclusion was that in a trivial case  you can comment the lines where you call the GPL libraries and ask the users/developers to un-comment them and take into consideration the GPL of the externals before using it. 

The version with the commented lines should be distributable under Apache, right? ;) :) 

Now scrolling through the discussion I see that you have made similar points also...

**Comment by nielsdrost** [comment]
> The version with the commented lines should be distributable under Apache, right? ;) :)

The smileys at the end of your sentence are exactly my point. You are doing tricks to be able to chance the license to "not GPL". Even if completely legal, it is not being nice to the author. It thus violates the "be nice" rule. And if you would like that in writing, see here: https://guide.esciencecenter.nl/software/documentation.html#code-of-conduct

**Comment by sonjageorgievska** [comment]
Hmm this sounds personal, although I know it was not your intention :). So, just to avoid confusion, I am not doing anything like this, just expressing an opinion on the topic... 

Niceness is another topic that I would not touch here. For example it is also nice to let everyone use the non-GPL part of the tool under Apache. Btw where exactly is the part on this topic in the link(s) on code of conduct? Then I will agree.

But anyhow, Zenodo allows you to assign a separate licence to every file included in your software.

**Comment by nielsdrost** [comment]
Ah, sorry, it was not my intention at all to judge the niceness of anyone!

My point is that our policy should reflect we are nice, and we should try to do "the right thing" not per-se the "legal thing".

- Niels

**Comment by LourensVeen** [comment]
Also, judges are humans, not computers, and they generally won't buy cutesy tricks like this. If a judge told you you were violating a license and ordered you to fix it, and you did it by commenting out some lines and changing the documentation to tell people to uncomment them, then I'd expect you to be held in contempt of court. (Which means you go to jail.)

Law is not a technical field, it's a humanities field. Treat it accordingly :-).

As for niceness, I also don't think it's nice to try to control what people do with code that they wrote themselves and which doesn't depend (any more) on your code, and I think it's nice if society decided that you don't need copyright permission from the implementor of an API to implement or use it yourself in your own code. So I'm happy with the outcome of Oracle vs. Google. And I still think dual-licensing your own source code (not the complete program) under GPL and Apache is fine, even if it has a GPL dependency.

**Comment by sonjageorgievska** [comment]
OK, I am out of this conversation now until indefinitely simply because I don't have time. But please do let us know if there is a consensus. And probably all of our colleagues that attended the 2015 workshop should be warned that some conclusions from it may not have been right...

**Comment by nielsdrost** [comment]
@LourensVeen, I believe we are getting close to consensus :-D

**Comment by RvanNieuwpoort** [comment]
Hi Niels, 

I fundamentally disagree with the comment you made on mar 16:

> My point is that most, if not all, authors use GPL with the intention of making sure all works that "build on" their work to also be GPL, and this is how everyone I know is both using and explaining the GPL. Case in point, the licence selection page by github puts it like this:
> 
> GNU GPLv3
> Permissions of this strong copyleft license are conditioned on making available complete source code of licensed works and modifications, which include larger works using a licensed work, under the same license. Copyright and license notices must be preserved. Contributors provide an express grant of patent rights.
> So, even if when reading the GPL carefully we do not (fully) agree with the summary above, this is the widely accepted interpretation of this license. We should respect the author's choice for this license, and not build on their work in a way they very likely did not want to happen.

I don't think you can project your own (or even widely accepted) intentions and interpretation of the GPL on others. No one can pretend to know what an author's intentions are with distributing the code under the GPL. This is pure speculation. Especially if these projected intentions clearly deviate from what the text literally says. This is IMHO morally questionable, and certainly legally unacceptable. We simply do not know what the author intentions behind using the GPL are. The only reasonable thing (morally and legally) is to take and interpret the license text as what is says. Even if your interpretation really is widely accepted (which I am not so sure about), this says nothing about the intentions of a specific author. 

On a side note, if this alternative interpretation really was the intended interpretation, the FSF would have simply fixed this in the GPLv3 (because this issue did not change between v2 and v3 as far as I know), or they could have released a GPLv4, right? 

My interpretation of "be nice" is not to project your own interpretation of a text onto others :-)  (NOTE THE SMILEY HERE !)

**Comment by bouweandela** [comment]
@LourensVeen I selected the parts of this PR that everyone agrees on and merged those because we want to move forward with the guide. Feel free to open an issue/make a new PR containing the rest.

**Pull request merged**